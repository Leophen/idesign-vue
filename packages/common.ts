import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
