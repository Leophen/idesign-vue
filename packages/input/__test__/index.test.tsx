import { vi } from 'vitest';
import { Input } from '../index';
import { render } from '../../common'

describe('Input 组件测试', () => {
  const renderInput = (input) => {
    const wrapper = render(input);
    return wrapper.find('.i-input__inner')
  }

  it('create', () => {
    const wrapper = render(<Input />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('init', () => {
    const changeFn = vi.fn();
    const wrapper = render(<Input placeholder='请输入内容' modelValue='iDesign' onInput={changeFn} />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('defaultValue', () => {
    const wrapper = render(<Input defaultValue='iDesign' />);
    expect(wrapper.element).toMatchSnapshot();
  });

  it('disabled', () => {
    const focusFn = vi.fn();
    const wrapper = render(<Input disabled onFocus={focusFn} />);
    const input = wrapper.find('.i-input')
    expect(input.classes()).toContain('i-input-is-disabled');
    input.trigger('click');
    expect(focusFn).toBeCalledTimes(0);
  });

  it('readonly', () => {
    const wrapper = render(<Input readonly />);
    expect(wrapper.find('.i-input').classes()).toContain('i-input-is-readonly');
  });

  it('size', () => {
    const wrapper = render(<Input size="small" />);
    expect(wrapper.find('.i-input').classes()).toContain('i-input--size-small');
  })

  it('maxLength', () => {
    const wrapper = render(<Input maxLength={15} />);
    expect(wrapper.find('.i-input--limit').exists()).toBe(true);
  })

  it('password', () => {
    const wrapper = render(<Input type="password" />);
    expect(wrapper.element).toMatchSnapshot();
  })

  it('number', () => {
    const wrapper = render(<Input type="number" />);
    expect(wrapper.element).toMatchSnapshot();
  })

  it('clearable', () => {
    const wrapper = render(<Input clearable />);
    expect(wrapper.element).toMatchSnapshot();
  })

  it('prefixIcon', () => {
    const wrapper = render(<Input prefixIcon="Search" />);
    expect(wrapper.find('.i-icon').classes()).toContain('i-input-prefix-icon');
  })

  it('suffixIcon', () => {
    const wrapper = render(<Input suffixIcon="Search" />);
    expect(wrapper.find('.i-icon').classes()).toContain('i-input-suffix-icon');
  })
});

