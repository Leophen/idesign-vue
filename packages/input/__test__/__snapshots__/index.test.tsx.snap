// Vitest Snapshot v1

exports[`Input 组件测试 > clearable 1`] = ``;

exports[`Input 组件测试 > create 1`] = ``;

exports[`Input 组件测试 > defaultValue 1`] = ``;

exports[`Input 组件测试 > init 1`] = ``;

exports[`Input 组件测试 > number 1`] = ``;

exports[`Input 组件测试 > password 1`] = ``;
